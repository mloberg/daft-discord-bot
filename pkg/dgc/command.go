package dgc

import "github.com/bwmarrin/discordgo"

// Command represents a Discord slash command
type Command struct {
	Name        string
	Description string
	Options     []*discordgo.ApplicationCommandOption

	Run func(s *discordgo.Session, i *discordgo.InteractionCreate) error
}
