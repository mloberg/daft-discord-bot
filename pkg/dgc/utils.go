package dgc

import "github.com/bwmarrin/discordgo"

var InteractionResponseEphemeral uint64 = 1 << 6

// GetUser returns the user that created the interaction either in a DM or guild
func GetUser(i *discordgo.InteractionCreate) *discordgo.User {
	if i.Member != nil {
		return i.Member.User
	}

	return i.User
}
