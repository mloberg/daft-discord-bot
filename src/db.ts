import { PrismaClient } from '@prisma/client';

const client = new PrismaClient({
    // TODO: setup logging
});

export default client;
