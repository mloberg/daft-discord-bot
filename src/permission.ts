import { Guild, GuildMember } from 'discord.js';

import config from './config';

export const ensureRole = async (guild: Guild): Promise<void> => {
    const role = guild.roles.cache.find((role) => role.name === config.djRole);
    if (role) {
        return;
    }

    await guild.roles.create({ data: { name: config.djRole } });
};
