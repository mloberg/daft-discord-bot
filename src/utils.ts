export function env(key: string, _default: string): string {
    return process.env[key] ?? _default;
}
