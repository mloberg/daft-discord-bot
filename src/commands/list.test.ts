import { Client, Message, TextChannel } from 'discord.js';
import { mocked } from 'ts-jest/utils';

import db from '../db';
import { FriendlyError } from '../error';
import { hasPermission } from '../permission';
import command from './list';

const mocks = {
    reply: jest.fn(),
    permission: mocked(hasPermission),
};

jest.mock('discord.js', () => ({
    Message: jest.fn().mockImplementation(() => ({
        reply: mocks.reply,
        member: { guild: { id: 'test' } },
    })),
}));

jest.mock('../permission');

describe('_play', () => {
    const client = {} as Client;
    const channel = {} as TextChannel;

    beforeEach(async () => {
        mocks.reply.mockClear();
        mocks.permission.mockClear();

        await db.song.create({
            data: {
                location: 'foo.mp3',
                guild: 'foo',
                tags: {
                    create: [{ tag: 'foo' }],
                },
            },
        });
        await db.song.create({
            data: {
                location: 'bar.mp3',
                guild: 'test',
                tags: {
                    create: [{ tag: 'bar' }],
                },
            },
        });
        await db.song.create({
            data: {
                location: 'test.mp3',
                guild: 'test',
                tags: {
                    create: [{ tag: 'foo' }, { tag: 'bar' }],
                },
            },
        });
    });

    afterEach(async () => {
        await db.$executeRaw`DELETE FROM tags`;
        await db.$executeRaw`DELETE FROM songs`;
        await db.$disconnect();
    });

    it('is a command', () => {
        expect(command.name).toBe('list');
        expect(command).toMatchSnapshot();
    });

    it('returns a list of available tags', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await command.run(message, { _: [], $0: 'list' });

        expect(mocks.reply).toHaveBeenCalledWith('bar (2), foo (1)');
    });

    it('will throw an error if no songs were found', async () => {
        await db.$executeRaw`DELETE FROM tags`;
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(true);

        await expect(command.run(message, { _: [], $0: 'list' })).rejects.toThrow(
            new FriendlyError('No tags found. Try adding some songs first.'),
        );
    });

    it('will throw an error if does not have role', async () => {
        const message = new Message(client, {}, channel);
        mocks.permission.mockReturnValue(false);

        await expect(command.run(message, { _: [], $0: 'list' })).rejects.toThrow(
            new FriendlyError('You do not have permission to do that.'),
        );
    });
});
