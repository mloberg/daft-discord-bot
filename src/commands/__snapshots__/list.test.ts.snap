// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_play is a command 1`] = `
Object {
  "description": "Show available song tags",
  "name": "list",
  "run": [Function],
}
`;
